### DISCUSSION TITLE
build: On linux, libadwaita installed using Homebrew is not found by zig build system

### DISCUSSION THREAD

**Discussion by lungsin** [ghostty-org/ghostty#3667] [Q&A] [answered]
In an attempt to package ghostty to homebrew, I tried to build ghostty by using the linux dependencies installed from homebrew:
```sh
brew install zig gtk4 libadwaita pkg-config
```

When running the build command `zig build -Doptimize=ReleaseFast`, I got the following error:
```sh
install
└─ install ghostty
   └─ zig build-exe ghostty ReleaseFast native failure
error: error: unable to find dynamic system library 'adwaita-1' using strategy 'mode_first'. searched paths:
  /home/linuxbrew/.linuxbrew/Cellar/gtk4/4.16.7/lib/libadwaita-1.so
  /home/linuxbrew/.linuxbrew/Cellar/pango/1.55.5/lib/libadwaita-1.so
  /home/linuxbrew/.linuxbrew/Cellar/harfbuzz/10.1.0/lib/libadwaita-1.so
  /home/linuxbrew/.linuxbrew/Cellar/gdk-pixbuf/2.42.12/lib/libadwaita-1.so
  /home/linuxbrew/.linuxbrew/Cellar/cairo/1.18.2/lib/libadwaita-1.so
  /home/linuxbrew/.linuxbrew/Cellar/graphene/1.10.8/lib/libadwaita-1.so
  /home/linuxbrew/.linuxbrew/Cellar/glib/2.82.4/lib/libadwaita-1.so
  /usr/lib/x86_64-linux-gnu/libadwaita-1.so
  /usr/local/lib64/libadwaita-1.so
  /usr/local/lib/libadwaita-1.so
  /usr/lib/x86_64-linux-gnu/libadwaita-1.so
  /lib64/libadwaita-1.so
  /lib/libadwaita-1.so
  /usr/lib64/libadwaita-1.so
  /usr/lib/libadwaita-1.so
  /lib/x86_64-linux-gnu/libadwaita-1.so
  /home/linuxbrew/.linuxbrew/Cellar/gtk4/4.16.7/lib/libadwaita-1.a
  /home/linuxbrew/.linuxbrew/Cellar/pango/1.55.5/lib/libadwaita-1.a
  /home/linuxbrew/.linuxbrew/Cellar/harfbuzz/10.1.0/lib/libadwaita-1.a
  /home/linuxbrew/.linuxbrew/Cellar/gdk-pixbuf/2.42.12/lib/libadwaita-1.a
  /home/linuxbrew/.linuxbrew/Cellar/cairo/1.18.2/lib/libadwaita-1.a
  /home/linuxbrew/.linuxbrew/Cellar/graphene/1.10.8/lib/libadwaita-1.a
  /home/linuxbrew/.linuxbrew/Cellar/glib/2.82.4/lib/libadwaita-1.a
  /usr/lib/x86_64-linux-gnu/libadwaita-1.a
  /usr/local/lib64/libadwaita-1.a
  /usr/local/lib/libadwaita-1.a
  /usr/lib/x86_64-linux-gnu/libadwaita-1.a
  /lib64/libadwaita-1.a
  /lib/libadwaita-1.a
  /usr/lib64/libadwaita-1.a
  /usr/lib/libadwaita-1.a
  /lib/x86_64-linux-gnu/libadwaita-1.a
```
The error message indicates that the zig build system was unable to find the `libadwaita` installed by homebrew, but it seems like it was able to find the lib gtk4 as indicated by the search paths printed by the error log: It included `/home/linuxbrew/.linuxbrew/Cellar/gtk4/4.16.7/lib/` as one of the search paths but not `/home/linuxbrew/.linuxbrew/Cellar/libadwaita/1.6.2/`.

I tried to dig the zig documentation, but wasn't able to find what `mode_first` strategy means. I want to ask the community that knows better about zig build system about how exactly zig determines the search paths for linux.

**Note**:  
- I'm using Pop OS 22.04, the build works as expected if the dependencies are installed using `apt` as instructed by the ghostty docs, so there's no problem with the intended way of building ghostty on Debian/Ubuntu. However, in order to be part of homebrew package, the dependencies need to be installed from homebrew as well.
- I'm using the v1.0.0 tagged source code from https://release.files.ghostty.org/1.0.0/ghostty-source.tar.gz. The build on main branch is also produces the same error.

**Comment by rhodes-b** [comment]
does `pkg-config` actually list gtk and libadwaita? as packages it found

**Reply by 00-kat** [reply]
To test, run `pkg-config --cflags --libs gtk4 libadwaita-1`.

Also consider trying `adwaita-1` if `libadwaita-1` doesn't work.

**Reply by 00-kat** [reply]
If `pkg-config` fails, it means you don't have a `.pc` file for libadwaita, which means that you don't have the developmental files for it.  I'm not familiar with how homebrew does things regarding this, though, so I can't help much more in that regard.

**Reply by lungsin** [reply]
I've just run the command you provided, yep looks like `pkg-config`  lists both gtk and libadwaita.
```sh
$ pkg-config --cflags --libs gtk4 libadwaita-1
-I/home/linuxbrew/.linuxbrew/Cellar/libadwaita/1.6.2/include/libadwaita-1 -I/home/linuxbrew/.linuxbrew/Cellar/gtk4/4.16.7/include/gtk-4.0 -I/home/linuxbrew/.linuxbrew/Cellar/pango/1.55.5/include/pango-1.0 -I/home/linuxbrew/.linuxbrew/Cellar/harfbuzz/10.1.0/include/harfbuzz -I/home/linuxbrew/.linuxbrew/Cellar/graphite2/1.3.14/include -I/home/linuxbrew/.linuxbrew/Cellar/gdk-pixbuf/2.42.12/include/gdk-pixbuf-2.0 -I/home/linuxbrew/.linuxbrew/opt/libtiff/include -I/home/linuxbrew/.linuxbrew/opt/zstd/include -I/home/linuxbrew/.linuxbrew/Cellar/xz/5.6.3/include -I/home/linuxbrew/.linuxbrew/opt/jpeg-turbo/include -I/home/linuxbrew/.linuxbrew/Cellar/cairo/1.18.2/include/cairo -I/home/linuxbrew/.linuxbrew/Cellar/fontconfig/2.15.0/include -I/home/linuxbrew/.linuxbrew/Cellar/expat/2.6.4/include -I/home/linuxbrew/.linuxbrew/opt/freetype/include/freetype2 -I/home/linuxbrew/.linuxbrew/opt/bzip2/include -I/home/linuxbrew/.linuxbrew/opt/libpng/include/libpng16 -I/home/linuxbrew/.linuxbrew/Cellar/libxext/1.3.6/include -I/home/linuxbrew/.linuxbrew/Cellar/xorgproto/2024.1/include -I/home/linuxbrew/.linuxbrew/Cellar/libxrender/0.9.12/include -I/home/linuxbrew/.linuxbrew/Cellar/libx11/1.8.10/include -I/home/linuxbrew/.linuxbrew/Cellar/libxcb/1.17.0/include -I/home/linuxbrew/.linuxbrew/Cellar/libxau/1.0.12/include -I/home/linuxbrew/.linuxbrew/Cellar/libxdmcp/1.1.5/include -I/home/linuxbrew/.linuxbrew/Cellar/pixman/0.44.2/include/pixman-1 -I/home/linuxbrew/.linuxbrew/Cellar/graphene/1.10.8/include/graphene-1.0 -I/home/linuxbrew/.linuxbrew/Cellar/graphene/1.10.8/lib/graphene-1.0/include -mfpmath=sse -msse -msse2 -I/home/linuxbrew/.linuxbrew/Cellar/fribidi/1.0.16/include/fribidi -I/home/linuxbrew/.linuxbrew/Cellar/appstream/1.0.4/include/appstream -I/home/linuxbrew/.linuxbrew/Cellar/glib/2.82.4/include -pthread -I/home/linuxbrew/.linuxbrew/opt/zlib/include -I/home/linuxbrew/.linuxbrew/Cellar/util-linux/2.40.2/include/libmount -I/home/linuxbrew/.linuxbrew/Cellar/util-linux/2.40.2/include/blkid -I/home/linuxbrew/.linuxbrew/Cellar/glib/2.82.4/include/glib-2.0 -I/home/linuxbrew/.linuxbrew/Cellar/glib/2.82.4/lib/glib-2.0/include -I/home/linuxbrew/.linuxbrew/Cellar/pcre2/10.44/include -I/home/linuxbrew/.linuxbrew/Cellar/libffi/3.4.6/include -L/home/linuxbrew/.linuxbrew/Cellar/libadwaita/1.6.2/lib -ladwaita-1 -L/home/linuxbrew/.linuxbrew/Cellar/gtk4/4.16.7/lib -lgtk-4 -L/home/linuxbrew/.linuxbrew/Cellar/pango/1.55.5/lib -lpangocairo-1.0 -lpango-1.0 -L/home/linuxbrew/.linuxbrew/Cellar/harfbuzz/10.1.0/lib -lharfbuzz -L/home/linuxbrew/.linuxbrew/Cellar/gdk-pixbuf/2.42.12/lib -lgdk_pixbuf-2.0 -L/home/linuxbrew/.linuxbrew/Cellar/cairo/1.18.2/lib -lcairo-gobject -lcairo -L/home/linuxbrew/.linuxbrew/Cellar/graphene/1.10.8/lib -lgraphene-1.0 -L/home/linuxbrew/.linuxbrew/Cellar/glib/2.82.4/lib -lgio-2.0 -lgobject-2.0 -lglib-2.0
```

**Reply by lungsin** [reply]
I made some progress on this issue. Here's what I found:

After reading ghostty's `build.zig`, looks like ghostty specifies `adwaita-1` as the library name to be linked ([permalink to the file](https://github.com/ghostty-org/ghostty/blob/06389b280acfdf0da0b2c8ea855582c5cedd1f8e/build.zig#L1429)). Also, the way zig 0.13.0 uses the pkg-config is by running `pkg-config --list-all` and do a couple of name matching strategy (exact match, appending `.0`, and trimming `-1.0` suffix) ([permalink to zig source code](https://github.com/ziglang/zig/blob/0.13.0/lib/std/Build/Step/Compile.zig#L654-L701)).

The problem is the libadwaita package uses `libadwaita-1.pc` to specify the pkg-config definition, meaning the name that pkg-config recognize is `libadwaita-1` and it mismatches with the ghostty specification `adwaita-1`. This is the case for the package from brew and apt. I confirmed that ghostty compiled successfully with the correct link library when replacing the `adwaita-1` to `libadwaita-1` in the source code.

I think the `libadwaita-1.pc` is something that's generated from the libadwaita project itself and not package manager specific. So, changing the `adwaita-1` to `libadwaita-1` should improve the compatibility with the `pkg-config` system. However, I'm not sure why it was specified as `adwaita-1` in the first place. It might be to integrate with something else I'm unaware of.

**Reply by 00-kat** [reply]
It's `libadwaita-1` on Arch Linux but somehow it seems to work just fine... 🤔

(`pkg-config --cflags --libs adwaita-1` doesn't work, but I'm able to compile Ghostty just fine...)

**Reply by lungsin** [reply]
Looks like zig fallbacks to linking by name directly if the pkg-config resolution fails ([zig source code](https://github.com/ziglang/zig/blob/0.13.0/lib/std/Build/Step/Compile.zig#L1151-L1157)). I think it compiles successfully because the compiler can find the `libadwaita-1` on the library path.

I think I'll mark this discussion as answered and open a new issue to discuss the renaming `adwaita-1` to `libadwaita-1`.

**Reply by jcollie** [reply]
Let Mitchell be the one to convert this to an issue. But I'd agree with switching to `libadwaita-1`, I check NixOS and they use `libadwaita-1.pc` as the name of the `pkg-config` file.

**Reply by lungsin** [reply]
> I think I'll mark this discussion as answered and open a new issue to discuss the renaming adwaita-1 to libadwaita-1

Oh wait, nvm, I've just read the contributing policy. Issues are supposed to be actionable. I guess we need to figure out first if it'll break something if `adwaita-1` is replaced with `libadwaita-1`.

**Reply by lungsin** [reply, marked as answer]
I did some thinking about what could broke if we replace the `adwaita-1` with `libadwaita-1` and I think it would affect linux system that doesn't use `pkg-config`. So I did a little experiment to remove `pkg-config` in my system (I did `sudo mv /usr/bin/pkg-config /usr/bin/pkg-config-backup`), and I got a build error that seems to complained that `pkg-config` was not found.

<details><summary>Error log</summary>
<p>

```sh
error: FileNotFound
/home/linuxbrew/.linuxbrew/Cellar/zig/0.13.0/lib/zig/std/process/Child.zig:423:5: 0x115c997 in waitPosix (build)
    return self.term.?;
    ^
/home/linuxbrew/.linuxbrew/Cellar/zig/0.13.0/lib/zig/std/process/Child.zig:300:9: 0x115ca13 in wait (build)
        try self.waitPosix();
        ^
/home/klungs/git/ghostty/build.zig:133:22: 0x1179d77 in build (build)
        const term = try pkgconfig.wait();
                     ^
/home/linuxbrew/.linuxbrew/Cellar/zig/0.13.0/lib/zig/std/Build.zig:2117:24: 0x113eb67 in runBuild__anon_8858 (build)
        .ErrorUnion => try build_zig.build(b),
                       ^
/home/linuxbrew/.linuxbrew/Cellar/zig/0.13.0/lib/zig/compiler/build_runner.zig:301:9: 0x1139e8f in main (build)
        try builder.runBuild(root);
```

</p>
</details> 

This indicates that ghostty with the current build config depends on `pkg-config` being installed on the system. This convinced me that it should be fine to rename `adwaita-1` to `libadwaita-1` since `pkg-config` is necessary anyway.

--- 
As a bonus, there's actually a config that can turn off pkg-config, and I'm curious about what would happen if this gets turn off, so I tried to modify the ghostty link config from the [build.zig](https://github.com/ghostty-org/ghostty/blob/e2f9eb6a6f4dc2108f91293938374c0ed314dcb8/build.zig#L1128-L1131) to disable pkg-config:
```zig
    const dynamic_link_opts: std.Build.Module.LinkSystemLibraryOptions = .{
        .preferred_link_mode = .dynamic,
        .search_strategy = .mode_first,
        .use_pkg_config = .no, // I tried to add this line to disable pkg-config
    };
```

Now, zig wasn't able to find the `gtk-4` library since the `.so` library installed from my apt is `libgtk-4.so` and the one searched by zig is `libgtk4.so` (without the dash `-` in the name).

<details><summary>Error log</summary>
<p>

```sh
install
└─ install ghostty
   └─ zig build-exe ghostty ReleaseFast native failure
error: error: unable to find dynamic system library 'gtk4' using strategy 'mode_first'. searched paths:
  /usr/lib/x86_64-linux-gnu/libgtk4.so
  /usr/local/lib64/libgtk4.so
  /usr/local/lib/libgtk4.so
  /usr/lib/x86_64-linux-gnu/libgtk4.so
  /lib64/libgtk4.so
  /lib/libgtk4.so
  /usr/lib64/libgtk4.so
  /usr/lib/libgtk4.so
  /lib/x86_64-linux-gnu/libgtk4.so
  /usr/lib/x86_64-linux-gnu/libgtk4.a
  /usr/local/lib64/libgtk4.a
  /usr/local/lib/libgtk4.a
  /usr/lib/x86_64-linux-gnu/libgtk4.a
  /lib64/libgtk4.a
  /lib/libgtk4.a
  /usr/lib64/libgtk4.a
  /usr/lib/libgtk4.a
  /lib/x86_64-linux-gnu/libgtk4.a
  ```

</p>
</details> 

So yeah, if in the future we want to support build without `pkg-config`, there is some refactoring work to be done to ensure that the library name is correct. But anyway, this is beyond the scope of this discussion and I don't think there's any need to support for now.

**Comment by daliusd** [comment]
I have used `brew install libadwaita` and it has worked for me (Ubuntu linux)

**Reply by lungsin** [reply]
Ah yup, the fix for this issue has been merged https://github.com/ghostty-org/ghostty/pull/4818

**Reply by daliusd** [reply]
Above command allowed to build, but run was failing. Following has helped:
`brew reinstall --build-from-source libadwaita`

Next time I must read https://ghostty.org/docs/install/build as well :)
